app/views/Project/EntryEdit/LeftPane/AssistPopup/utils.ts

```diff
@@ -0,0 +1,232 @@
+    ScaleMappingsItem,
+    ScaleWidget,
+    SelectMappingsItem,
+    SingleSelectWidget,
+    MultiSelectMappingsItem,
+    MultiSelectWidget,
+} from '#types/newAnalyticalFramework';
+import {
+} from '#types/newEntry';
+
+import {
+    getType,
+    DeepReplace,
+} from '#utils/types';
+
+import {
+    PartialAttributeType,
+} from '../../schema';
+
+type Matrix1dWidgetAttribute = getType<PartialAttributeType, { widgetType: 'MATRIX1D' }>;
+type Matrix2dWidgetAttribute = getType<PartialAttributeType, { widgetType: 'MATRIX2D' }>;
+type ScaleWidgetAttribute = getType<PartialAttributeType, { widgetType: 'SCALE' }>;
+type SingleSelectWidgetAttribute = getType<PartialAttributeType, { widgetType: 'SELECT' }>;
+type MultiSelectWidgetAttribute = getType<PartialAttributeType, { widgetType: 'MULTISELECT' }>;
+
+export function createMatrix1dAttr(
+    mappings: Matrix1dMappingsItem[] | undefined,
+    widget: Matrix1dWidget,
+): Matrix1dWidgetAttribute | undefined {
+    if (!mappings || mappings.length < 1) {
+        return undefined;
+    }
+    const mappingsGroupedByRows = listToGroupList(
+        mappings,
+        (m) => m.association.rowKey,
+        (m) => m.association.subRowKey,
+    );
+
+    const value = mapToMap(
+        mappingsGroupedByRows,
+        (m) => m,
+        (d) => listToMap(d, (k) => k, () => true),
+    );
+
+    return {
+        clientId: randomString(),
+        widget: widget.id,
+        widgetVersion: widget.version,
+        widgetType: 'MATRIX1D',
+        data: { value },
+    };
+}
+
+interface SubRowMap {
+    type: 'SUB_ROW';
+    rowKey: string;
+    subRowKey: string;
+}
+
+type ColumnMap = {
+    type: 'COLUMN';
+    columnKey: string;
+} | {
+    type: 'SUB_COLUMN';
+    columnKey: string;
+    subColumnKey: string;
+};
+
+export function filterColumn(
+    mappingsItem: Matrix2dMappingsItem,
+): mappingsItem is DeepReplace<Matrix2dMappingsItem, SubRowMap, never> {
+    return mappingsItem.association.type === 'COLUMN' || mappingsItem.association.type === 'SUB_COLUMN';
+}
+
+export function filterSubRows(
+    mappingsItem: Matrix2dMappingsItem,
+): mappingsItem is DeepReplace<Matrix2dMappingsItem, ColumnMap, never> {
+    return mappingsItem.association.type === 'SUB_ROW';
+}
+
+export function createMatrix2dAttr(
+    mappings: Matrix2dMappingsItem[] | undefined,
+    widget: Matrix2dWidget,
+): Matrix2dWidgetAttribute | undefined {
+    if (!mappings || mappings.length < 1) {
+        return undefined;
+    }
+    const columns = mappings.filter(filterColumn);
+    const rows = mappings.filter(filterSubRows);
+
+    if (columns.length === 0 || rows.length === 0) {
+        return undefined;
+    }
+
+    const groupedCols = listToGroupList(
+        columns,
+        (col) => col.association.columnKey,
+        (col) => col.association.subColumnKey,
+    );
+
+    const transformedCols = mapToMap(
+        groupedCols,
+        (colKey) => colKey,
+        (subColumns) => subColumns.filter(isDefined),
```

Comment:
Why do we need to filter subColumns? Is the col.association.subColumnKey not always defined?